Feature/maskings (#8006)
arangodb/arangodb: Documentation/Books/Manual/Programs/Arangodump/Maskings.md

      ...
    }

> Using `"*"` has collection name, defines a default behavior for collections not

Author: done

document. See below for a complete description of the parameters of
"xify_front".

> ### masking vs. dump-data

Author: done

*arangodump* also supports a very coarse masking with the option

> `--dump-data false`.  This basically removed all data from the dump.

Author: Done

You can either use `--masking` or `--dump-data false`, but not both.

### masking vs. include-collection

*arangodump* also suppers a very coarse masking with the option
`--include-collection`. This will restrict the collections that are
dumped.

It is possible to combine `--masking` and `--include-collection`

> this will take the intersection of expertable collections.

Author: Done

Path
----

If the path starts with a `.` then it is considered to be a prefix wildcard
match.  For example, `.name` will match the attribute name `name` all leaf

> attributes in the document. Leaf attributes are attributes of type `string`,

Author: done

> `number`, `bool` and `array` (see below). `name` will only match leaf attributes

Author: added null

    "path": "`name.with.dots`"

If the attribute value is an array the masking is applied to all the

> array elements.

Author: done

> ### Example

Reviewer: Headlines must be unique per document. The easiest way to fix is to use bold text instead:

```suggestion
**Example**
```

Author: done

> This following will replace name by a "XXXX" masked string

Author: Done

    {
      "type": "xify_front",
      "path": ".name",
      "unmaskedLength": 2

>     }

Reviewer: This should rather be using fenced code (in triple backticks, with the first four spaces removed) and JSON highlighting

```
```json
{
  ...
}
```

Author: done

> The object

Reviewer: ```suggestion
The document:
```

Author: done

    {
      "name": "top-level-name",
      "age": 42,
      "nicknames" : [ { "name": "hugo" }, "egon" ],
      "other": {
        "name": [ "emil", { "secret": "superman" } ]
      }
    }

> will be replaced as follows:

Reviewer: ```suggestion
… will be changed as follows:
```

Author: done

    {
      "name": "xxxxxxxxxxxxme",
      "age": 42,
      "nicknames" : [ { "name": "xxgo" }, "egon" ],
      "other": {
        "name": [ "xxil", { "secret": "superman" } ]
      }

>       }

Author: Done

> `egon` and `superman` are not replaced, because they are not in an

Author: done

> attribute value of an attribute `name`.

Reviewer: ```suggestion
attribute value of which the attribute name is `name`.
```

Author: done

> ### sub-objects and arrays

Reviewer: ```suggestion
### Nested objects and arrays
```

Author: done

If you specify a path and the attribute value is an array then the

> masking decision is applied to each entry of this array as if this

Reviewer: ```suggestion
masking decision is applied to each element of the array as if this
```

Author: done

> were the value of the attribute.

Reviewer: ```suggestion
was the value of the attribute.
```

Author: done

> If the attribute value is a object, then the attribute is not masked.

Reviewer: ```suggestion
If the attribute value is an object, then the attribute is not masked.
```

Author: done

Instead the sub-object is checked further for leaf attributes.

> #### Example

Reviewer: ```suggestion
**Example**
```

Author: done

> Masking `email` will convert

Reviewer: ```suggestion
Masking `email` will convert:
```

Author: done

  { 
    "email" : "email address" 
  }

> into

Reviewer: ```suggestion
… into:
```

Author: done

  { 
    "email" : "xxil xxxxxxss" 
  }

> because `email` is a leaf sttribute.

Reviewer: ```suggestion
… because `email` is a leaf attribute.
```

Author: @Simran-B Do you mean "put three dots in front"?

Reviewer: It's an ellipsis symbol. I like the following style:

```
For example:

    console.log("foo");

… does this and that.
```

The colon indicates that what follows belongs to this sentence. And the ellipsis signifies that the sentence continues after the code example.

Author: understood

  { 
    "email" : [ 
      "address one", 
      "address two" 
    ] 
  } 

> will be converted into

Reviewer: ```suggestion
… will be converted into:
```

Author: done

  { 
    "email" : [ 
      "xxxxxss xne", 
      "xxxxxss xwo" 
    ] 
  } 

> because the array is "unfolded".

Reviewer: ```suggestion
… because the array is "unfolded".
```

Author: @Simran-B I might have misunderstood the "..."? Do your really mean put "..." into the text?

Reviewer: I meant to write these dots literally, see my [other comment](https://github.com/arangodb/arangodb/pull/8006#discussion_r)

  { 
    "email" : { 
      "address" : "email address" 

>     } 

Author: done

    }

> will not be changed because `email` is not a leaf attribute.

Reviewer: ```suggestion
… will not be changed because `email` is not a leaf attribute.
```

Masking Functions
-----------------

### random_string

> This masking replaces character strings with a random string of the

Reviewer: I assume it applies to all kinds of strings? _Character strings_ makes it sound a bit like it would only replace strings with letters, but not number string (_key) for instance.

Author: done

> same length. Other values, e.g. numbers and booleans, are not changed.

Reviewer: Here it says "of same length", but further below the example configuration it says "anonymized string" and that the length is not guaranteed to be the same?!

We should add a full negative list here (which data types are not changed). But random_string includes strings in arrays right?

Would the result of masking this with _random_string_:
`[null, false, true, 0.5, 123, "foobar", [ "baz" ], { "sub": "abcdef" }`
… be as follows?
`[null, false, true, 0.5, 123, "xPgalY", [ "mqW" ], { "sub": "abcdef" }`

Author: done

    {
      "path": ".name",
      "type": "random_string"
    }

> This will replace all values of attributes with key `name` with a

Author: done

anonymized string. It is not guaranteed that the string will be of

> the same length.

Author: done

A hash of the original string is computed. If the original string is

> shorted then the hash will be used. This will result in a longer

Author: Done

> string. If the string is longer than hash than the hash characters

Author: done

will be repeated as many times as needed to reach the full original

> string length.

Reviewer: I am confused. The behavior is still not clear to me, because there are some statements which apparently contradict each other. Is the following correct?
- A hash is computed from the input string (constant length?)
- If the computed hash is longer than the input string then the hash value is trimmed to the same length (first n characters?), e.g. input `abc` and hash `Xiaabo`, then the output would be `Xia`
- If the computed hash is shorter than the input string then the hash value is repeated, but limited to the length of the input string (e.g. if the input is `abcdefg` and the hash `Xia`, then the output would be `XiaXiaX`, not `XiaXiaXia`)

Author: Sorry for the confusion. String can get longer. Fixed and added example

The random string will use the characters `a` to `z` and `A`to `Z`.

    {
      "path": ".name",
      "type": "random_string",
      "characterSet": "abcd01234"
    }

In this case the random string will use the characters `a`, `b`, `c`,
`d` as well as the digits `0`, `1`, `2`, `3`, `4`. If `characterSet`
is empty, it defaults to `a` to `z` and `A` to `Z`.

> ### xify_front (Enterprise)

Reviewer: Is _xify_front_ a common name for this?

Is this meant for enterprises, or is it only available in the Enterprise Edition?

Author: done

> each run. This avoids dictionary attacks.

Reviewer: Should we say why this is done? (To prevent that someone figures out the original value from the hash)

Author: done

> If you need reproducable results, i.e. hash that do not change between

Author: done

different runs of *arangodump*, you need to specify a secret as seed,
a number which must not be `0`.

    {
      "type": "xify_front",
      "path": ".name",
      "unmaskedLength": 2,
      "hash": true,
      "seed": 246781478647
    }

### zip (Enterprise)

This masking replaces a zip code with a random one. If the attribute value

> is not a string it is replaced by `12345`.

Reviewer: Replaced by a number or a string `"12345"`?

    {
      "path": ".code",
      "type": "zip",
    }

> This will replace an existing zip with a random one. It uses the following

Reviewer: ```suggestion
This will replace a real zip code with a random one. It uses the following
```

Author: done

rule: If a character of the original zip code is a digit it will be replaced
by a random digit. If a character of the original zip code is a letter it
will be replaced by a random letter keeping the case.

    {
      "path": ".code",
      "type": "zip",
      "default": "abcdef"
    }

> If the attribute value is not a string use the value of default `abcdef`.

Reviewer: This is already stated above, can we remove it in one of the places?

> If the attribute value is not a string use the value of default `abcdef`.

Author: done

> #### Example

Reviewer: ```suggestion
**Example**
```

> If the original zip code is

Reviewer: ```suggestion
If the original zip code is:
```

Author: done

    50674

> it will be replaced by (for example)

Reviewer: ```suggestion
… it will be replaced by e.g.
```

Author: done

    98146

> If the original zip code is

Reviewer: ```suggestion
If the original zip code is:
```

Author: done

    SA34-EA

> it will be replaced by (for example)

Reviewer: ```suggestion
… it will be replaced by e.g.
```

Author: done

    OW91-JI

> Please note that this will generate random zip. Therefore there is a chance

Reviewer: ```suggestion
Note that this will generate random zip code. Therefore there is a chance
```

Author: done

> Please note that this will generate random zip. Therefore there is a chance

Author: done

that it will destroy uniqueness.

### date-time (Enterprise)

This masking replaces the value of the attribute with a random date.

    {
      "type": "date-time",
      "begin" : "2019-01-01",
      "end": "2019-12-31",
      "format": "%yyyy-%mm-%dd",
    }

`begin` and `end` are in ISO8601.

The format is described in

> [DATE_FORMAT](https://docs.arangodb.com/3.4/AQL/Functions/Date.html#dateformat).

Reviewer: ```suggestion
[DATE_FORMAT](../../../AQL/Functions/Date.html#dateformat).
```

Author: Done

> ### integral number (Enterprise)

Reviewer: ```suggestion
### Integral number (Enterprise)
```

Author: done

> This masking replaces the value of the attribute with a random integral number.

Reviewer: Always, or only if the attribute type is number?

Author: done

    {
      "type": "integer",
      "lower" : -100,
      "upper": 100
    }

> ### decimal (Enterprise)

Reviewer: ```suggestion
### Decimal (Enterprise)
```

Author: done

This masking replaces the value of the attribute with a random decimal.

    {
      "type": "float",
      "lower" : -0.3,
      "upper": 0.3
    }

> by default, the decimal has a scale of 2.

Reviewer: ```suggestion
By default, the decimal has a scale of 2.
```

Author: done

    {
      "type": "float",
      "lower" : -0.3,
      "upper": 0.3,
      "scale": 3
    }

> will use a scale of 3.

Reviewer: ```suggestion
… will use a scale of 3.
```

Author: done

> will use a scale of 3.

Author: done

> ### credit card number (Enterprise)

Reviewer: ```suggestion
### Credit card number (Enterprise)
```

Author: done

This masking replaces the value of the attribute with a random credit card number.

    {
      "type": "ccnumber",
    }

See [Luhn](https://en.wikipedia.org/wiki/Luhn_algorithm) for details.

> ### phone number (Enterprise)

Reviewer: ```suggestion
### Phone number (Enterprise)
```

Author: done

This masking replaces a phone number with a random one. If the attribute value

> is not a string it is replaced by `+1234567890`.

Reviewer: `12345` as number or string `"12345"`?

> is not a string it is replaced by `+1234567890`.

Author: I have changed this to `"+1234567890"` and added the word string

    {
      "type": "phone",
      "default": "+4912345123456789"
    }

This will replace an existing phone number with a random one. It uses
the following rule: If a character of the original number is a digit

> it will be replaced by a random digit. If it is letter it is replaced

Author: done

by a letter. All other characters are unchanged.

    {
      "type": "zip",
      "default": "+4912345123456789"
    }

If the attribute value is not a string use the value of default

> `+4912345123456789`.

Author: added quotes

> ### email address (Enterprise)

Reviewer: ```suggestion
### Email address (Enterprise)
```

Author: done

> This takes the email the email address, computes a hash value and

Reviewer: ```suggestion
This takes an email address, computes a hash value and
```

Author: done

> split this is three equal parts `AAAA`, `BBBB`, and `CCCC`. The

Author: done
